### PR TITLE
Storing and retrieving data using merkle trees

### PR DESCRIPTION
**Pull request by tbekas**
Work in progress. Done so far:
* storing and retrieving data using merkle trees works locally
* merkle trees are stored as regular blocks

To do:
* fix all TODOs
* fix block exchange - it uses old mechanism that doesn't work anymore

### CODE REVIEW

File: codex/stores/repostore.nim [hunk 1]
```diff
@@ -97,7 +102,53 @@ method getBlock*(self: RepoStore, cid: Cid): Future[?!Block] {.async.} =
     return failure(newException(BlockNotFoundError, err.msg))
 
   trace "Got block for cid", cid
-  return Block.new(cid, data)
+  return Block.new(cid, data, verify = true)
+
+proc getMerkleTree(self: RepoStore, cid: Cid): Future[?!MerkleTree] {.async.} =
+  try:
```
Comment (dryajov) [inline codex/stores/repostore.nim:108]: I'd suggest avoiding exceptions (specially mixing wiht result), you can either use `.catch` from `questionable` (?) or the `withValue` apis (this are ugly but performant).

File: codex/blocktype.nim [hunk 2]
```diff
@@ -147,24 +147,25 @@ func new*(
     cid: cid,
     data: @data).success
 
-func new*(
+proc new*(
```
Comment (dryajov) [inline codex/blocktype.nim:150]: Any reason to change to a proc, func has no side effects and are overall "safer", I don't see anything that would trigger a side effect here?

File: codex/stores/repostore.nim (hunk 1 above)
Comment (tbekas) [reply codex/stores/repostore.nim:108]: Fixed in #566 , I'm using `.getOption` on `LruCache` instead

File: codex/blocktype.nim (hunk 2 above)
Comment (tbekas) [reply codex/blocktype.nim:181]: Fixed in #566 , it's `func` once again

**Pull request closed without merging**